src/adb.js

```diff
@@ -1 +1,393 @@
+		}
+
+		if (typeof appId !== 'string' || !appId) {
+			return Promise.reject(new TypeError('Expected app ID to be a string.'));
+		}
+
+		return this
+			.shell(deviceId, 'ps')
+			.then(data => {
+				if (data) {
+					const lines = data.toString().split('\n');
+					for (let i = 0, j = lines.length; i < j; i++ ) {
+						let columns = lines[i].trim().split(/\s+/);
+						if (columns.pop() === appId) {
+							return parseInt(columns[1]);
+						}
+					}
+				}
+				return 0;
+			});
+	}
+
+	/**
+	 * Installs an app to the specified device/emulator.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} apkFile - The application apk file to install.
+	 * @param {Object} [opts] - Install options.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	installApp(deviceId, apkFile, opts = {}) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		apkFile = util.expandPath(apkFile);
+		if (!util.existsSync(apkFile)) {
+			return Promise.reject(new TypeError(`APK file "${apkFile}" does not exist.`));
+		}
+
+		return this
+			.devices()
+			.then(devices => {
+				if (!devices.some(d => d.id === deviceId)) {
+					return Promise.reject(new Error('Device not found.'));
+				}
+			})
+			.then(sdk.detect)
+			.then(result => util.run(result.sdk.executables.adb, ['-s', deviceId, 'install', '-r', apkFile]));
+	}
+
+	/**
+	 * Starts an application on the specified device/emulator.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} appId - The application's id.
+	 * @param {String} activity - The name of the activity to run.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	startApp(deviceId, appId, activity) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		if (typeof appId !== 'string' || !appId) {
+			return Promise.reject(new TypeError('Expected app ID to be a string.'));
+		}
+
+		if (typeof activity !== 'string' || !activity) {
+			return Promise.reject(new TypeError('Expected activity name to be a string.'));
+		}
+
+		return this.shell(deviceId, `am start -n ${appId} /. ${activity.replace(/^\./, '')}`);
+	}
+
+	/**
+	 * Stops an application on the specified device/emulator.
+	 *
+	 * @param {String} deviceId - The id of the device or emulator.
+	 * @param {String} appId - The application's id.
+	 * @returns {Promise}
+	 * @access public
+	 */
+	stopApp(deviceId, appId) {
+		if (typeof deviceId !== 'string' || !deviceId) {
+			return Promise.reject(new TypeError('Expected device ID to be a string.'));
+		}
+
+		if (typeof appId !== 'string' || !appId) {
+			return Promise.reject(new TypeError('Expected app ID to be a string.'));
+		}
+
+		return this
+			.getPid(deviceId, appId)
+			.then(pid => this.shell(deviceId, `am force-stop ${appId}`));
```

Comment:
Need to make sure that `pid` is not zero. Need to check if result from `am force-stop` is `Unknown command: force-stop`, and if so do another shell command and `kill <pid>`.